hpat/_str_ext.cpp

```diff
@@ -311,6 +316,68 @@ void* str_from_float64(double in)
     return new std::string(std::to_string(in));
 }
 
+#if PY_VERSION_HEX >= 0x03000000
+#define PyString_Check(name) PyUnicode_Check(name)
+#define PyString_AsString(str) PyUnicode_AsUTF8(str)
+#endif
+
+/// @brief create a concatenated string and offset table from a pandas series of strings
+/// @note strings in returned buffer will not be 0-terminated.
+/// @param[out] buffer newly allocated buffer with concatenated strings, or NULL
+/// @param[out] no_strings number of strings concatenated, value < 0 indicates an error
+/// @param[out] offset_table newly allocated array of no_strings+1 integers
+///                          first no_strings entries denote offsets, last entry indicates size of output array
+/// @param[in]  obj Python Sequence object, intended to be a pandas series of string
+void string_array_from_sequence(PyObject * obj, int64_t * no_strings, uint32_t ** offset_table, char ** buffer)
+{
+#define CHECK(expr, msg) if(!(expr)){std::cerr << msg << std::endl; PyGILState_Release(gilstate); return;}
+
+    auto gilstate = PyGILState_Ensure();
+
+    *no_strings = -1;
+    *offset_table = NULL;
+    *buffer = NULL;
+
+    CHECK(PySequence_Check(obj), "expecting a PySequence");
+    CHECK(no_strings && offset_table && buffer, "output arguments must not be NULL");
+
+    Py_ssize_t n = PyObject_Size(obj);
+    if(n == 0 ) {
+        // empty sequence, this is not an error, need to set size
+        PyGILState_Release(gilstate);
+        no_strings = 0;
+        return;
+    }
+
+    uint32_t * offsets = new uint32_t[n];
+    std::vector<const char *> tmp_store(n);
+    size_t len = 0;
+    for(Py_ssize_t i = 0; i < n; ++i) {
+        offsets[i] = len;
+        PyObject * s = PySequence_GetItem(obj, i);
+        CHECK(s, "getting element failed");
+        CHECK(PyString_Check(s), "expecting a string");
+        tmp_store[i] = PyString_AsString(s);
```

Comment:
@fschlimb I think we need to deallocate the strings returned from `PyUnicode_AsUTF8`

https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsUTF8AndSize

Reply:
the doc says: "The caller is not responsible for deallocating the buffer."
We release the reference to the PyObject in 363.

Reply:
Sorry, I misread the page :)